### ISSUE TITLE
Too many temporary files after tests

### ISSUE THREAD

**Issue by g4s8** [zerocracy/farm#435]
Unit tests spawn a lot of temporary files which are not removed after tests. It may produce few Gb in a day. I think `File.deleteOnExit()` method for `FkItem` may help.

**Comment by 0crat** [comment]
The job [#435](https://github.com/zerocracy/farm/issues/435) assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/611443849). The budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by amihaiemil** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/zerocracy/farm/issues/435#issuecomment-370693095))

@amihaiemil You can't do that, unless you have one of these roles: `ARC`, `PO`. Your current roles are: `DEV`.

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/zerocracy/farm/issues/435#issuecomment-370693095))

@amihaiemil The impediment for [#435](https://github.com/zerocracy/farm/issues/435) was registered successfully by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by amihaiemil** [comment]
@g4s8 I looked into it. ``FkItem`` does have the ``delete`` boolean attribute according to which the file is deleted or not in ``close()``. 

However, this attribute is initialized in the ctor and there is also a ctor which sets it to ``false``:

https://github.com/zerocracy/farm/blob/91513b6161574d9c4cda293fa6123afa6d3b736f/src/main/java/com/zerocracy/farm/fake/FkItem.java#L64

Now, modifying that ctor is not OK, since its used a lot, in many places (including tests and prod code). I assume ``FkItem`` is used indirectly in tests via ``FkProject`` -- so I wanted to change the initialization there and pass ``true``... however, ``FkProject`` is also used in many places inside prod code. 

![image](https://user-images.githubusercontent.com/6305156/37210502-9f9473e4-23b1-11e8-8bf7-6fac6b820e59.png)

Bottom line is, I think we have a bigger issue here: these fake objects (which should primarly be used for testing) are also used in the prod codebase. I would refactor this and use ``Null Objects`` in prod code, not these fake objects. 

Then, once the logic is separated, we can decide which fake/null items to keep and which not. What do you think? I would close this ticket and open another one for the suggested refactoring.

**Comment by amihaiemil** [comment]
@0crat waiting for answers

**Comment by 0crat** [comment]
> @0crat waiting for answers ([here](https://github.com/zerocracy/farm/issues/435#issuecomment-371923350))

@amihaiemil Job [#435](https://github.com/zerocracy/farm/issues/435) is already on hold

**Comment by amihaiemil** [comment]
@g4s8 ping, can you answer my question?

**Comment by g4s8** [comment]
@amihaiemil sorry, missed the notification.

You are right, it's not clear why fake object are used in production, but I think it may be OK to use `/tmp` files here (it depends on logic), also I think it will be expected behavior to loose temporary files even in production after restart. So I would add a `todo` to get rid of fake objects here, but put `deleteOnExit(true)` also.

**Comment by amihaiemil** [comment]
@g4s8 But the problem is that ``close()`` method on ``FkItem`` may be called sooner, not just when shutting down the app (for instance, if the instnace is ina try-with-resources) -- what if it's called sooner? I have no idea what is the logic where it is used and what could break. I think there was a solid reason for that ``delete`` flag in ``FkItem``...

**Comment by amihaiemil** [comment]
@g4s8 to be honest, I don't even understand why we need files on disk at all :D data stored in AWS S3 (or Mongo, as I saw somewhere) -- why put it on disk after reading from there?

**Comment by g4s8** [comment]
@amihaiemil 
> why put it on disk after reading from there?

maybe to cache it and pay less to Amazon for S3?

@yegor256 could you please comment [here](https://github.com/zerocracy/farm/issues/435#issuecomment-372984441):
> I have no idea what is the logic where it is used and what could break. I think there was a solid reason for that delete flag in FkItem...

**Comment by yegor256** [comment]
@amihaiemil @g4s8 S3Item is caching files on disc, in order to work faster. If we remove that, we will have thousands of requests to AWS -- this would be slow and expensive (they charge per each request).

The `delete` flag in `FkItem` is there mostly in order to clean up the disc. But in some rare cases it's not necessary, since we want the file to remain there.

**Comment by amihaiemil** [comment]
@yegor256 
>  But in some rare cases it's not necessary, since we want the file to remain there.

Bingo, that's what I figured.

But if it's an ``FkItem``, a fake object, why does it need to create a real file on disk anyway? Couldn't it work all in memory?

**Comment by yegor256** [comment]
@amihaiemil well, `Item` has method `path()`, which returns a real physical path of the file on the disc. That's why we need to work with files :)

**Comment by amihaiemil** [comment]
@0crat refuse

**Comment by 0crat** [comment]
Tasks refusal is discouraged, see [§6](http://www.zerocracy.com/policy.html#6): -15 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
The job [#435](https://github.com/zerocracy/farm/issues/435) assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/971224370); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html)

**Comment by carlosmiranda** [comment]
@g4s8 see PR #435

**Comment by carlosmiranda** [comment]
@0crat wait for merge of #913

**Comment by 0crat** [comment]
> @0crat wait for merge of #913 ([here](https://github.com/zerocracy/farm/issues/435#issuecomment-379965667))

@carlosmiranda The impediment for [#435](https://github.com/zerocracy/farm/issues/435) was registered successfully by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by carlosmiranda** [comment]
@g4s8 PR #913 closed, can we close this now?

**Comment by g4s8** [comment]
@carlosmiranda thanks! Now I can switch to tmpfs back ;)

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#435](https://github.com/zerocracy/farm/issues/435) is now out of scope

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>All impediments registered in the ticket have explanations of the reasons and links to related tickets, if necessary

Please add details to registsred [impediment](https://github.com/zerocracy/farm/issues/435#issuecomment-370693095).

**Comment by carlosmiranda** [comment]
@ypshenychka job has been refused by @amihaiemil and continued by me. Why does he have to provide details?

**Comment by amihaiemil** [comment]
@ypshenychka updated my comment, I was waiting for clarifications

**Comment by ypshenychka** [comment]
@carlosmiranda comment was written by @amihaiemil. You could ask him to make corrections prior to QA review.

**Comment by ypshenychka** [comment]
@amihaiemil Thank you!

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +30 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)
